nested SSR pages (and API routes) are throwing a module not found error (needs additional patch for adapters)

## Summary
Haven't tested for production yet, but at least when running `greenwood develop` and having a nested SSR page route, e.g.
```
src/
  pages/
    blog/
      index.js
```

Will output
```sh
Started local development server at http://localhost:1984
      Develop command with HTML route response using default export and nested SSR page
Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/Users/owenbuckley/Workspace/project-evergreen/greenwood/packages/cli/test/cases/develop.ssr/src/pages/index.js' imported from /Users/owenbuckley/Workspace/project-evergreen/greenwood/packages/cli/src/lib/execute-route-module.js
    at new NodeError (node:internal/errors:405:5)
    at finalizeResolution (node:internal/modules/esm/resolve:327:11)
    at moduleResolve (node:internal/modules/esm/resolve:980:10)
    at defaultResolve (node:internal/modules/esm/resolve:1193:11)
    at ModuleLoader.defaultResolve (node:internal/modules/esm/loader:404:12)
    at ModuleLoader.resolve (node:internal/modules/esm/loader:373:25)
    at ModuleLoader.getModuleJob (node:internal/modules/esm/loader:250:38)
    at ModuleLoader.import (node:internal/modules/esm/loader:336:34)
    at importModuleDynamically (node:internal/modules/esm/translators:143:35)
    at importModuleDynamicallyCallback (node:internal/modules/esm/utils:176:14) {
  url: 'file:///Users/owenbuckley/Workspace/project-evergreen/greenwood/packages/cli/test/cases/develop.ssr/src/pages/index.js',
  code: 'ERR_MODULE_NOT_FOUND'
```

## Details
If you notice in the path shown in the error, the _blog/_ segment is missing
```sh
Cannot find module '/Users/owenbuckley/Workspace/project-evergreen/greenwood/packages/cli/test/cases/develop.ssr/src/pages/index.js' 
```

--- a/packages/cli/src/config/rollup.config.js
+++ b/packages/cli/src/config/rollup.config.js
@@ -274,20 +274,29 @@ function greenwoodImportMetaUrl(compilation) {
 // TODO could we use this instead?
 // https://github.com/rollup/rollup/blob/v2.79.1/docs/05-plugin-development.md#resolveimportmeta
 // https://github.com/ProjectEvergreen/greenwood/issues/1087
-function greenwoodPatchSsrPagesEntryPointRuntimeImport() {
+function greenwoodPatchSsrPagesEntryPointRuntimeImport(compilation) {
   return {
     name: 'greenwood-patch-ssr-pages-entry-point-runtime-import',
     generateBundle(options, bundle) {
       Object.keys(bundle).forEach((key) => {
-        if (key.startsWith('__')) {
-          // ___GWD_ENTRY_FILE_URL=${filename}___
+        // map rollup bundle names back to original SSR pages for output bundles and paths
+        if (key.startsWith('_')) {
           const needle = bundle[key].code.match(/___GWD_ENTRY_FILE_URL=(.*.)___/);
-          if (needle) {
+
+          if (bundle[key].facadeModuleId.startsWith(compilation.context.scratchDir.pathname) && needle) {
             const entryPathMatch = needle[1];
 
-            bundle[key].code = bundle[key].code.replace(/'___GWD_ENTRY_FILE_URL=(.*.)___'/, `new URL('./_${entryPathMatch}', import.meta.url)`);
-          } else {
-            console.warn(`Could not find entry path match for bundle => ${key}`);
+            Object.keys(bundle).forEach((_) => {
+              if (bundle[_].facadeModuleId === `${compilation.context.pagesDir.pathname}${entryPathMatch}`) {
+                bundle[key].code = bundle[key].code.replace(/'___GWD_ENTRY_FILE_URL=(.*.)___'/, `new URL('./${bundle[_].fileName}', import.meta.url)`);
+
+                compilation.graph.forEach((page, idx) => {
+                  if (page.relativeWorkspacePagePath === `/${entryPathMatch}`) {
+                    compilation.graph[idx].outputPath = key;
+                  }
+                });
+              }
+            });
           }
         }
       });
@@ -405,7 +414,7 @@ const getRollupConfigForSsr = async (compilation, input) => {
       }),
       commonjs(),
       greenwoodImportMetaUrl(compilation),
-      greenwoodPatchSsrPagesEntryPointRuntimeImport() // TODO a little hacky but works for now
+      greenwoodPatchSsrPagesEntryPointRuntimeImport(compilation) // TODO a little hacky but works for now
     ],
     onwarn: (errorObj) => {
       const { code, message } = errorObj;

--- a/packages/cli/src/config/rollup.config.js
+++ b/packages/cli/src/config/rollup.config.js
@@ -278,16 +278,20 @@ function greenwoodPatchSsrPagesEntryPointRuntimeImport(compilation) {
   return {
     name: 'greenwood-patch-ssr-pages-entry-point-runtime-import',
     generateBundle(options, bundle) {
+      const { pagesDir, scratchDir } = compilation.context;
+
       Object.keys(bundle).forEach((key) => {
         // map rollup bundle names back to original SSR pages for output bundles and paths
         if (key.startsWith('_')) {
           const needle = bundle[key].code.match(/___GWD_ENTRY_FILE_URL=(.*.)___/);
 
-          if (bundle[key].facadeModuleId.startsWith(compilation.context.scratchDir.pathname) && needle) {
+          // handle windows shenanigans for facadeModuleId and path separators
+          if (new URL(`file://${bundle[key].facadeModuleId}`).pathname.startsWith(scratchDir.pathname) && needle) {
             const entryPathMatch = needle[1];
 
             Object.keys(bundle).forEach((_) => {
-              if (bundle[_].facadeModuleId === `${compilation.context.pagesDir.pathname}${entryPathMatch}`) {
+              // handle windows shenanigans for facadeModuleId and path separators
+              if (new URL(`file://${bundle[_].facadeModuleId}`).pathname === `${pagesDir.pathname}${entryPathMatch}`) {
                 bundle[key].code = bundle[key].code.replace(/'___GWD_ENTRY_FILE_URL=(.*.)___'/, `new URL('./${bundle[_].fileName}', import.meta.url)`);
 
                 compilation.graph.forEach((page, idx) => {

--- a/packages/cli/src/lifecycles/bundle.js
+++ b/packages/cli/src/lifecycles/bundle.js
@@ -190,9 +190,10 @@ async function bundleSsrPages(compilation) {
 
     for (const page of compilation.graph) {
       if (page.isSSR && !page.prerender) {
-        const { filename, imports, route, template, title, relativeWorkspacePagePath } = page;
+        const { imports, route, template, title, relativeWorkspacePagePath } = page;
         const entryFileUrl = new URL(`./_${relativeWorkspacePagePath.replace('/', '')}`, scratchDir);
         const moduleUrl = new URL(`./${relativeWorkspacePagePath.replace('/', '')}`, pagesDir);
+        const outputPathRootUrl = new URL(`file://${path.dirname(entryFileUrl.pathname)}`);
         const request = new Request(moduleUrl); // TODO not really sure how to best no-op this?
         // TODO getTemplate has to be static (for now?)
         // https://github.com/ProjectEvergreen/greenwood/issues/955
@@ -205,8 +206,13 @@ async function bundleSsrPages(compilation) {
         staticHtml = await (await htmlOptimizer.optimize(new URL(`http://localhost:8080${route}`), new Response(staticHtml))).text();
         staticHtml = staticHtml.replace(/[`\\$]/g, '\\$&'); // https://stackoverflow.com/a/75688937/417806
 
+        if (!await checkResourceExists(outputPathRootUrl)) {
+          await fs.mkdir(outputPathRootUrl, {
+            recursive: true
+          });
+        }
+
         // better way to write out this inline code?
-        await fs.mkdir(entryFileUrl.pathname.replace(filename, ''), { recursive: true });
         await fs.writeFile(entryFileUrl, `
           import { executeRouteModule } from '${normalizePathnameForWindows(executeModuleUrl)}';
 

--- a/packages/cli/src/lifecycles/bundle.js
+++ b/packages/cli/src/lifecycles/bundle.js
@@ -190,9 +190,9 @@ async function bundleSsrPages(compilation) {
 
     for (const page of compilation.graph) {
       if (page.isSSR && !page.prerender) {
-        const { filename, imports, route, template, title } = page;
-        const entryFileUrl = new URL(`./_${filename}`, scratchDir);
-        const moduleUrl = new URL(`./${filename}`, pagesDir);
+        const { filename, imports, route, template, title, relativeWorkspacePagePath } = page;
+        const entryFileUrl = new URL(`./_${relativeWorkspacePagePath.replace('/', '')}`, scratchDir);
+        const moduleUrl = new URL(`./${relativeWorkspacePagePath.replace('/', '')}`, pagesDir);
         const request = new Request(moduleUrl); // TODO not really sure how to best no-op this?
         // TODO getTemplate has to be static (for now?)
         // https://github.com/ProjectEvergreen/greenwood/issues/955
@@ -206,13 +206,14 @@ async function bundleSsrPages(compilation) {
         staticHtml = staticHtml.replace(/[`\\$]/g, '\\$&'); // https://stackoverflow.com/a/75688937/417806
 
         // better way to write out this inline code?
+        await fs.mkdir(entryFileUrl.pathname.replace(filename, ''), { recursive: true });
         await fs.writeFile(entryFileUrl, `
           import { executeRouteModule } from '${normalizePathnameForWindows(executeModuleUrl)}';
 
           export async function handler(request) {
             const compilation = JSON.parse('${JSON.stringify(compilation)}');
             const page = JSON.parse('${JSON.stringify(page)}');
-            const moduleUrl = '___GWD_ENTRY_FILE_URL=${filename}___';
+            const moduleUrl = '___GWD_ENTRY_FILE_URL=${relativeWorkspacePagePath.replace('/', '')}___';
             const data = await executeRouteModule({ moduleUrl, compilation, page, request });
             let staticHtml = \`${staticHtml}\`;
 

--- a/packages/cli/src/lifecycles/graph.js
+++ b/packages/cli/src/lifecycles/graph.js
@@ -191,6 +191,7 @@ const generateGraph = async (compilation) => {
              * data: custom page frontmatter
              * filename: base filename of the page
              * id: filename without the extension
+             * relativeWorkspacePagePath: the file path relative to the user's workspace directory
              * label: "pretty" text representation of the filename
              * imports: per page JS or CSS file imports to be included in HTML output from frontmatter
              * resources: sum of all resources for the entire page
@@ -206,6 +207,7 @@ const generateGraph = async (compilation) => {
               data: customData || {},
               filename,
               id,
+              relativeWorkspacePagePath: relativePagePath,
               label: id.split('-')
                 .map((idPart) => {
                   return `${idPart.charAt(0).toUpperCase()}${idPart.substring(1)}`;

--- a/packages/cli/src/lifecycles/serve.js
+++ b/packages/cli/src/lifecycles/serve.js
@@ -294,7 +294,7 @@ async function getHybridServer(compilation) {
       const request = transformKoaRequestIntoStandardRequest(url, ctx.request);
 
       if (!config.prerender && matchingRoute.isSSR && !matchingRoute.prerender) {
-        const { handler } = await import(new URL(`./__${matchingRoute.filename}`, outputDir));
+        const { handler } = await import(new URL(`./${matchingRoute.outputPath}`, outputDir));
         const response = await handler(request, compilation);
 
         ctx.body = Readable.from(response.body);

--- a/packages/cli/src/plugins/resource/plugin-standard-html.js
+++ b/packages/cli/src/plugins/resource/plugin-standard-html.js
@@ -105,7 +105,7 @@ class StandardHtmlResource extends ResourceInterface {
     }
 
     if (matchingRoute.isSSR) {
-      const routeModuleLocationUrl = new URL(`./${matchingRoute.filename}`, pagesDir);
+      const routeModuleLocationUrl = new URL(`.${matchingRoute.relativeWorkspacePagePath}`, pagesDir);
       const routeWorkerUrl = this.compilation.config.plugins.find(plugin => plugin.type === 'renderer').provider().executeModuleUrl;
 
       await new Promise(async (resolve, reject) => {

--- a/packages/cli/test/cases/build.plugins.context/theme-pack-context-plugin.js
+++ b/packages/cli/test/cases/build.plugins.context/theme-pack-context-plugin.js
@@ -1,20 +1,15 @@
 import fs from 'fs/promises';
-import os from 'os';
-import { spawnSync } from 'child_process';
 
 const packageJson = JSON.parse(await fs.readFile(new URL('./package.json', import.meta.url), 'utf-8'));
-const myThemePackPlugin = () => [{
+const myThemePackPlugin = (options = {}) => [{
   type: 'context',
   name: 'my-theme-pack:context',
   provider: () => {
     const { name } = packageJson;
-    const command = os.platform() === 'win32' ? 'npm.cmd' : 'npm';
-    const ls = spawnSync(command, ['ls', name]);
 
-    const isInstalled = ls.stdout.toString().indexOf('(empty)') < 0;
-    const templateLocation = isInstalled
-      ? new URL(`./node_modules/${name}/dist/layouts/`, import.meta.url)
-      : new URL('./fixtures/layouts/', import.meta.url);
+    const templateLocation = options.__isDevelopment // eslint-disable-line no-underscore-dangle
+      ? new URL('./fixtures/layouts/', import.meta.url)
+      : new URL(`./node_modules/${name}/dist/layouts/`, import.meta.url);
 
     return {
       templates: [

--- a/packages/cli/test/cases/develop.plugins.context/greenwood.config.js
+++ b/packages/cli/test/cases/develop.plugins.context/greenwood.config.js
@@ -24,7 +24,9 @@ class MyThemePackDevelopmentResource extends ResourceInterface {
 
 export default {
   plugins: [
-    ...myThemePackPlugin(),
+    ...myThemePackPlugin({
+      __isDevelopment: true
+    }),
     {
       type: 'resource',
       name: 'my-theme-pack:resource',

--- a/packages/cli/test/cases/develop.ssr/develop.ssr.spec.js
+++ b/packages/cli/test/cases/develop.ssr/develop.ssr.spec.js
@@ -16,6 +16,9 @@
  *   components/
  *     footer.js
  *   pages/
+ *     blog
+ *       first-post.js
+ *       index.js
  *     artists.js
  *     post.js
  *   templates/
@@ -282,6 +285,94 @@ describe('Develop Greenwood With: ', function() {
 
         expect(paragraph.length).to.equal(1);
         expect(paragraph[0].textContent).to.not.be.undefined;
+      });
+    });
+
+    describe('Develop command with HTML route response using default export and nested SSR Blog Index page', function() {
+      let response = {};
+      let dom = {};
+      let body;
+
+      before(async function() {
+        response = await fetch(`${hostname}/blog/`);
+        body = await response.clone().text();
+        dom = new JSDOM(body);
+      });
+
+      it('should return a 200 status', function(done) {
+        expect(response.status).to.equal(200);
+        done();
+      });
+
+      it('should return the correct content type', function(done) {
+        expect(response.headers.get('content-type')).to.equal('text/html');
+        done();
+      });
+
+      it('should return a response body', function(done) {
+        expect(body).to.not.be.undefined;
+        done();
+      });
+
+      it('should be valid HTML from JSDOM', function(done) {
+        expect(dom).to.not.be.undefined;
+        done();
+      });
+
+      it('should be valid HTML from JSDOM', function(done) {
+        expect(dom).to.not.be.undefined;
+        done();
+      });
+
+      it('should have the expected postId as an <h1> tag in the body', function() {
+        const heading = dom.window.document.querySelectorAll('body > h1');
+
+        expect(heading.length).to.equal(1);
+        expect(heading[0].textContent).to.equal('Nested SSR page should work!');
+      });
+    });
+
+    describe('Develop command with HTML route response using default export and nested SSR Blog First Post page', function() {
+      let response = {};
+      let dom = {};
+      let body;
+
+      before(async function() {
+        response = await fetch(`${hostname}/blog/first-post/`);
+        body = await response.clone().text();
+        dom = new JSDOM(body);
+      });
+
+      it('should return a 200 status', function(done) {
+        expect(response.status).to.equal(200);
+        done();
+      });
+
+      it('should return the correct content type', function(done) {
+        expect(response.headers.get('content-type')).to.equal('text/html');
+        done();
+      });
+
+      it('should return a response body', function(done) {
+        expect(body).to.not.be.undefined;
+        done();
+      });
+
+      it('should be valid HTML from JSDOM', function(done) {
+        expect(dom).to.not.be.undefined;
+        done();
+      });
+
+      it('should be valid HTML from JSDOM', function(done) {
+        expect(dom).to.not.be.undefined;
+        done();
+      });
+
+      it('should have the expected postId as an <h1> tag in the body', function() {
+        const heading = dom.window.document.querySelectorAll('body > h1');
+
+        expect(heading.length).to.equal(1);
+        expect(heading[0].textContent).to.equal('Nested SSR First Post page should work!');
       });
     });
   });

--- a/packages/cli/test/cases/develop.ssr/src/pages/blog/first-post.js
+++ b/packages/cli/test/cases/develop.ssr/src/pages/blog/first-post.js
@@ -1,0 +1,7 @@
+export default class BlogFirstPostPage extends HTMLElement {
+  connectedCallback() {
+    this.innerHTML = `
+      <h1>Nested SSR First Post page should work!</h1>
+    `;
+  }
+}

--- a/packages/cli/test/cases/develop.ssr/src/pages/blog/index.js
+++ b/packages/cli/test/cases/develop.ssr/src/pages/blog/index.js
@@ -1,0 +1,7 @@
+export default class BlogPage extends HTMLElement {
+  connectedCallback() {
+    this.innerHTML = `
+      <h1>Nested SSR page should work!</h1>
+    `;
+  }
+}

--- a/packages/cli/test/cases/serve.default.ssr/serve.default.ssr.spec.js
+++ b/packages/cli/test/cases/serve.default.ssr/serve.default.ssr.spec.js
@@ -23,6 +23,9 @@
  *   pages/
  *     about.md
  *     artists.js
+ *     blog/
+ *       first-post.js
+ *       index.js
  *     index.js
  *     post.js
  *     users.js
@@ -269,6 +272,94 @@ describe('Serve Greenwood With: ', function() {
           .filter(file => file.indexOf('users.js') >= 0);
 
         expect(scriptFiles.length).to.equal(2);
+      });
+    });
+
+    describe('Serve command with HTML route response using default export and nested SSR Blog Index page', function() {
+      let response = {};
+      let dom = {};
+      let body;
+
+      before(async function() {
+        response = await fetch(`${hostname}/blog/`);
+        body = await response.clone().text();
+        dom = new JSDOM(body);
+      });
+
+      it('should return a 200 status', function(done) {
+        expect(response.status).to.equal(200);
+        done();
+      });
+
+      it('should return the correct content type', function(done) {
+        expect(response.headers.get('content-type')).to.equal('text/html');
+        done();
+      });
+
+      it('should return a response body', function(done) {
+        expect(body).to.not.be.undefined;
+        done();
+      });
+
+      it('should be valid HTML from JSDOM', function(done) {
+        expect(dom).to.not.be.undefined;
+        done();
+      });
+
+      it('should be valid HTML from JSDOM', function(done) {
+        expect(dom).to.not.be.undefined;
+        done();
+      });
+
+      it('should have the expected postId as an <h1> tag in the body', function() {
+        const heading = dom.window.document.querySelectorAll('body > h1');
+
+        expect(heading.length).to.equal(1);
+        expect(heading[0].textContent).to.equal('Nested SSR page should work!');
+      });
+    });
+
+    describe('Develop command with HTML route response using default export and nested SSR Blog First Post page', function() {
+      let response = {};
+      let dom = {};
+      let body;
+
+      before(async function() {
+        response = await fetch(`${hostname}/blog/first-post/`);
+        body = await response.clone().text();
+        dom = new JSDOM(body);
+      });
+
+      it('should return a 200 status', function(done) {
+        expect(response.status).to.equal(200);
+        done();
+      });
+
+      it('should return the correct content type', function(done) {
+        expect(response.headers.get('content-type')).to.equal('text/html');
+        done();
+      });
+
+      it('should return a response body', function(done) {
+        expect(body).to.not.be.undefined;
+        done();
+      });
+
+      it('should be valid HTML from JSDOM', function(done) {
+        expect(dom).to.not.be.undefined;
+        done();
+      });
+
+      it('should be valid HTML from JSDOM', function(done) {
+        expect(dom).to.not.be.undefined;
+        done();
+      });
+
+      it('should have the expected postId as an <h1> tag in the body', function() {
+        const heading = dom.window.document.querySelectorAll('body > h1');
+
+        expect(heading.length).to.equal(1);
+        expect(heading[0].textContent).to.equal('Nested SSR First Post page should work!');
       });
     });
 

--- a/packages/cli/test/cases/serve.default.ssr/src/pages/blog/first-post.js
+++ b/packages/cli/test/cases/serve.default.ssr/src/pages/blog/first-post.js
@@ -1,0 +1,7 @@
+export default class BlogFirstPostPage extends HTMLElement {
+  connectedCallback() {
+    this.innerHTML = `
+      <h1>Nested SSR First Post page should work!</h1>
+    `;
+  }
+}

--- a/packages/cli/test/cases/serve.default.ssr/src/pages/blog/index.js
+++ b/packages/cli/test/cases/serve.default.ssr/src/pages/blog/index.js
@@ -1,0 +1,7 @@
+export default class BlogPage extends HTMLElement {
+  connectedCallback() {
+    this.innerHTML = `
+      <h1>Nested SSR page should work!</h1>
+    `;
+  }
+}

--- a/www/pages/docs/data.md
+++ b/www/pages/docs/data.md
@@ -62,6 +62,8 @@ graph {
 
   outputPath, // (string) the relative path to write to when generating static HTML
 
+  relativeWorkspacePagePath, // the file path relative to the user's workspace directory
+
   path, // (string) path to the file
 
   route,  // (string) A URL, typically derived from the filesystem path, e.g. /blog/2019/first-post/

